File: solution/day20/src/test/java/original/YahtzeeCalculatorTests.java
package original;

import builders.DiceBuilder;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static domain.yahtzee.original.YahtzeeCalculator.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class YahtzeeCalculatorTests {
    public static Stream<Arguments> numbers() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(1, 2, 1, 1, 3), 1, 3),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 6), 1, 0),
                Arguments.of(DiceBuilder.newRoll(4, 4, 4, 4, 4), 1, 0),
                Arguments.of(DiceBuilder.newRoll(4, 1, 4, 4, 5), 4, 12)
        );
    }

    @MethodSource("numbers")
    @ParameterizedTest
    void sum_for_numbers_for(DiceBuilder dice, int number, int expectedResult) {
        assertThat(number(dice.build(), number))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> threeOfAKinds() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(3, 3, 3, 4, 5), 18),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 6), 0),
                Arguments.of(DiceBuilder.newRoll(4, 4, 4, 4, 4), 20),
                Arguments.of(DiceBuilder.newRoll(1, 1, 2, 1, 5), 10)
        );
    }

    @MethodSource("threeOfAKinds")
    @ParameterizedTest
    void three_of_a_kind_for(DiceBuilder dice, int expectedResult) {
        assertThat(threeOfAKind(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> fourOfAKinds() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(3, 3, 3, 3, 5), 17),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 6), 0),
                Arguments.of(DiceBuilder.newRoll(4, 4, 4, 4, 4), 20),
                Arguments.of(DiceBuilder.newRoll(1, 1, 1, 1, 5), 9)
        );
    }

    @MethodSource("fourOfAKinds")
    @ParameterizedTest
    void four_of_a_kind_for(DiceBuilder dice, int expectedResult) {
        assertThat(fourOfAKind(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> fullHouses() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(2, 2, 3, 3, 3), 25),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 6), 0),
                Arguments.of(DiceBuilder.newRoll(4, 4, 1, 4, 1), 25)
        );
    }

    @MethodSource("fullHouses")
    @ParameterizedTest
    void full_house_for(DiceBuilder dice, int expectedResult) {
        assertThat(fullHouse(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> smallStraights() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(1, 2, 3, 4, 5), 30),
                Arguments.of(DiceBuilder.newRoll(5, 4, 3, 2, 1), 30),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 1), 30),
                Arguments.of(DiceBuilder.newRoll(1, 1, 1, 3, 2), 0)
        );
    }

    @MethodSource("smallStraights")
    @ParameterizedTest
    void small_straight_for(DiceBuilder dice, int expectedResult) {
        assertThat(smallStraight(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> largeStraights() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(1, 2, 3, 4, 5), 40),
                Arguments.of(DiceBuilder.newRoll(5, 4, 3, 2, 1), 40),
                Arguments.of(DiceBuilder.newRoll(2, 3, 4, 5, 6), 40),
                Arguments.of(DiceBuilder.newRoll(1, 4, 1, 3, 2), 0)
        );
    }

    @MethodSource("largeStraights")
    @ParameterizedTest
    void large_straight_for(DiceBuilder dice, int expectedResult) {
        assertThat(largeStraight(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> yahtzees() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(4, 4, 4, 4, 4), 50),
                Arguments.of(DiceBuilder.newRoll(2, 2, 2, 2, 2), 50),
                Arguments.of(DiceBuilder.newRoll(1, 4, 1, 3, 2), 0)
        );
    }

    @MethodSource("yahtzees")
    @ParameterizedTest
    void yahtzeeFor(DiceBuilder dice, int expectedResult) {
        assertThat(yahtzee(dice.build()))
                .isEqualTo(expectedResult);
    }

    public static Stream<Arguments> chances() {
        return Stream.of(
                Arguments.of(DiceBuilder.newRoll(3, 3, 3, 3, 3), 15),
                Arguments.of(DiceBuilder.newRoll(6, 5, 4, 3, 3), 21),
                Arguments.of(DiceBuilder.newRoll(1, 4, 1, 3, 2), 11)
        );
    }

    @MethodSource("chances")
    @ParameterizedTest
    void chanceFor(DiceBuilder dice, int expectedResult) {
        assertThat(chance(dice.build()))
                .isEqualTo(expectedResult);
    }

    @Nested
    class FailFor {
        public static Stream<Arguments> invalidRollLengths() {
            return Stream.of(
                    Arguments.of(new int[]{1}),
                    Arguments.of(new int[]{1, 1}),
                    Arguments.of(new int[]{1, 6, 2}),
                    Arguments.of(new int[]{1, 6, 2, 5}),
                    Arguments.of(new int[]{1, 6, 2, 5, 4, 1}),
                    Arguments.of(new int[]{1, 6, 2, 5, 4, 1, 2})
            );
        }

        @MethodSource("invalidRollLengths")
        @ParameterizedTest
        void invalid_roll_lengths(int... dice) {
            assertThatThrownBy(() -> number(dice, 1)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> threeOfAKind(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> fourOfAKind(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> fullHouse(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> smallStraight(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> largeStraight(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> yahtzee(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> chance(dice)).isInstanceOf(IllegalArgumentException.class);
        }

        public static Stream<Arguments> invalidDieInRolls() {
            return Stream.of(
                    Arguments.of(new int[]{1, 1, 1, 1, 7}),
                    Arguments.of(new int[]{0, 1, 1, 1, 2}),
                    Arguments.of(new int[]{1, 1, -1, 1, 2})
            );
        }

        @MethodSource("invalidDieInRolls")
        @ParameterizedTest
        void invalid_die_in_rolls(int[] dice) {
            assertThatThrownBy(() -> number(dice, 1)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> threeOfAKind(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> fourOfAKind(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> fullHouse(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> smallStraight(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> largeStraight(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> yahtzee(dice)).isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> chance(dice)).isInstanceOf(IllegalArgumentException.class);
        }
    }
}

File: solution/kotlin/day20/src/test/kotlin/original/YahtzeeCalculatorTests.kt
package original

import builders.DiceBuilder.Companion.`🎲🎲🎲🎲🎲`
import domain.yahtzee.original.YahtzeeCalculator
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe
import parameters.InvalidRoll
import parameters.NumberRollCase
import parameters.RollCase

class YahtzeeCalculatorTests : FunSpec({
    context("add only numbers for numbers") {
        withData(
            NumberRollCase(`🎲🎲🎲🎲🎲`(1, 2, 1, 1, 3), 1, 3),
            NumberRollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 6), 1, 0),
            NumberRollCase(`🎲🎲🎲🎲🎲`(4, 4, 4, 4, 4), 1, 0),
            NumberRollCase(`🎲🎲🎲🎲🎲`(4, 1, 4, 4, 5), 4, 12)
        ) { (dice, number, expectedResult) ->
            YahtzeeCalculator.number(dice.build(), number) shouldBe expectedResult
        }
    }

    context("sum of all dice for 3 of a kind") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(3, 3, 3, 4, 5), 18),
            RollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 6), 0),
            RollCase(`🎲🎲🎲🎲🎲`(4, 4, 4, 4, 4), 20),
            RollCase(`🎲🎲🎲🎲🎲`(1, 1, 2, 1, 5), 10)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.threeOfAKind(dice.build()) shouldBe expectedResult
        }
    }

    context("sum of all dice for 4 of a kind") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(3, 3, 3, 3, 5), 17),
            RollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 6), 0),
            RollCase(`🎲🎲🎲🎲🎲`(4, 4, 4, 4, 4), 20),
            RollCase(`🎲🎲🎲🎲🎲`(1, 1, 1, 1, 5), 9)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.fourOfAKind(dice.build()) shouldBe expectedResult
        }
    }

    context("25 for full houses") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(2, 2, 3, 3, 3), 25),
            RollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 6), 0),
            RollCase(`🎲🎲🎲🎲🎲`(4, 4, 1, 4, 1), 25)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.fullHouse(dice.build()) shouldBe expectedResult
        }
    }

    context("30 for small straights") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(1, 2, 3, 4, 5), 30),
            RollCase(`🎲🎲🎲🎲🎲`(5, 4, 3, 2, 1), 30),
            RollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 1), 30),
            RollCase(`🎲🎲🎲🎲🎲`(1, 1, 1, 3, 2), 0)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.smallStraight(dice.build()) shouldBe expectedResult
        }
    }

    context("40 for large straights") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(1, 2, 3, 4, 5), 40),
            RollCase(`🎲🎲🎲🎲🎲`(5, 4, 3, 2, 1), 40),
            RollCase(`🎲🎲🎲🎲🎲`(2, 3, 4, 5, 6), 40),
            RollCase(`🎲🎲🎲🎲🎲`(1, 4, 1, 3, 2), 0)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.largeStraight(dice.build()) shouldBe expectedResult
        }
    }

    context("50 for yahtzees") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(4, 4, 4, 4, 4), 50),
            RollCase(`🎲🎲🎲🎲🎲`(2, 2, 2, 2, 2), 50),
            RollCase(`🎲🎲🎲🎲🎲`(1, 4, 1, 3, 2), 0)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.yahtzee(dice.build()) shouldBe expectedResult
        }
    }

    context("sum of all dice for chance") {
        withData(
            RollCase(`🎲🎲🎲🎲🎲`(3, 3, 3, 3, 3), 15),
            RollCase(`🎲🎲🎲🎲🎲`(6, 5, 4, 3, 3), 21),
            RollCase(`🎲🎲🎲🎲🎲`(1, 4, 1, 3, 2), 11)
        ) { (dice, expectedResult) ->
            YahtzeeCalculator.chance(dice.build()) shouldBe expectedResult
        }
    }

    context("fail for invalid rolls") {
        withData(
            InvalidRoll(1),
            InvalidRoll(1, 6, 2),
            InvalidRoll(1, 6, 2, 5),
            InvalidRoll(1, 6, 2, 5, 4, 1),
            InvalidRoll(1, 6, 2, 5, 4, 1, 2)
        ) {
            it.dice
                .let {
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.number(it, 1) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.threeOfAKind(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.fourOfAKind(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.fullHouse(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.smallStraight(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.largeStraight(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.yahtzee(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.chance(it) }
                }
        }
    }

    context("fail for invalid die in roll") {
        withData(
            InvalidRoll(1, 1, 1, 1, 7),
            InvalidRoll(0, 1, 1, 1, 2),
            InvalidRoll(1, 1, -1, 1, 2)
        ) {
            it.dice
                .let {
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.number(it, 1) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.threeOfAKind(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.fourOfAKind(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.fullHouse(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.smallStraight(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.largeStraight(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.yahtzee(it) }
                    shouldThrow<IllegalArgumentException> { YahtzeeCalculator.chance(it) }
                }
        }
    }
})